### PR TITLE
Implement multi field sort

### LINKED ISSUE #2: RFC: JSON API Multiple Field Sorting
## Update 2018-03: fully implemented

---

_From @bravo-kernel on April 13, 2017 18:52_

The JsonApiListener currently does not fully support [JSON API Sorting](http://jsonapi.org/format/#fetching-sorting) functionality:

- [X] single field sorting (works)
- [ ] multi field sorting (not working)

> A server MAY choose to support requests to sort resource collections according to one or more criteria (“sort fields”).

```
GET /people?sort=age,name HTTP/1.1
Accept: application/vnd.api+json
```

_Copied from original issue: FriendsOfCake/crud#522_

### PR DESCRIPTION
**Pull request by rchavik**
- [x] multi field sorting
- [x] sorting contained associations, eg: /uses?include=address&sort=address.street
- [x] tests
- [x] direction
- [x] pagination results

Closes #2

### CODE REVIEW

File: src/Listener/PaginationListener.php [hunk 1]
```diff
@@ -74,6 +74,11 @@ protected function _getJsonApiPaginationResponse(array $pagination)
     {
         $routerMethod = 'normalize'; // produce relative links
 
+        $defaultUrl = array_intersect_key($pagination, [
```
Comment (josegonzalez) [inline src/Listener/PaginationListener.php:77]: Can you explain this change?

File: src/Listener/JsonApiListener.php [hunk 2]
```diff
@@ -341,6 +344,62 @@ public function beforeFind(Event $event)
...
+                if (!in_array($include, $includes)) {
+                    continue;
+                }
+
+                $associations = $repository->associations();
+                foreach ($associations as $association) {
+                    $subject->query->contain([
+                        $association->alias() => [
+                            'sort' => [
+                                $association->aliasField($field) => $direction,
```
Comment (josegonzalez) [inline src/Listener/JsonApiListener.php]: Are we sorting _every_ table by the same field here?

Comment (rchavik) [reply src/Listener/JsonApiListener.php]: Hmm, you're right. i'll get it fixed

File: src/Listener/PaginationListener.php (hunk 1 above)
Comment (rchavik) [reply src/Listener/PaginationListener.php:77]: we pull the `page` and `sort` configuration to be later re-used when constructing the pagination links:

eg: 
https://github.com/FriendsOfCake/crud-json-api/pull/29/files/eb40b3eda90b61475df4fbecf689e185df8cac0f#diff-a7c2bafd4fbe2043ab51dd175cf91e41R8

**Comment by josegonzalez** [comment]
@rchavik is it possible to manage sorts for associations or no?

File: src/Listener/JsonApiListener.php (hunk 2 above)
Comment (rchavik) [reply src/Listener/JsonApiListener.php]: Thanks 

Bug is proven in d46c8f3, fixed in ae044cc

**Comment by rchavik** [comment]
> @rchavik is it possible to manage sorts for associations or no?

not sure what you mean by manage here.

but you can specify `/currencies?include=countries&sort=-countries.code` to sort by countries code descending.

**Comment by rchavik** [comment]
I think this is good to go.

File: src/Listener/JsonApiListener.php [hunk 3]
```diff
@@ -341,6 +344,65 @@ public function beforeFind(Event $event)
...
+     */
+    protected function _sortParameter($sortFields, Subject $subject, $options)
+    {
+        if (is_string($sortFields)) {
+            $sortFields = explode(',', $sortFields);
+        }
+        $sortFields = array_filter((array)$sortFields);
+
+        $order = [];
+        $includes = $this->config('include');
```
Comment (josegonzalez) [inline src/Listener/JsonApiListener.php:364]: Can you explain the `include` functionality?

**Comment by josegonzalez** [comment]
Does this need anymore documentation on usage?

**Comment by bravo-kernel** [comment]
@josegonzalez @rchavik If you can add a new (empty/placeholder) page named `Sorting` to the documentation under `Configuration` that will suffice for now. I will then push an additional PR with the docs myself this weekend. 

@rchavik after you answer the latest question on the `include` I will merge, looks good to me

File: src/Listener/JsonApiListener.php (hunk 3 above)
Comment (rchavik) [reply src/Listener/JsonApiListener.php:364]: As far as I know, this is an existing functionality, processed by `_includeParameter` in  ff033cb (Current version [here](https://github.com/FriendsOfCake/crud-json-api/blob/1fe01b0849d4074ddf3c6a38e7cb20890d24c1d5/src/Listener/JsonApiListener.php#L288). You basically pass an comma separated values of valid types that is related to the type in context, eg:

`/countries?include=currencies,national_cities`

The resulting response should include the currencies + national cities in the payload.

From reading the code,  those values are read from the request and set accordingly in https://github.com/FriendsOfCake/crud-json-api/blob/1fe01b0849d4074ddf3c6a38e7cb20890d24c1d5/src/Listener/JsonApiListener.php#L311

Comment (bravo-kernel) [reply src/Listener/JsonApiListener.php:364]: That makes sense, I think that would be part of the functionality added by mr. @dakota.

**Comment by bravo-kernel** [comment]
@josegonzalez does this PR have your blessing now?

**Review by josegonzalez — APPROVED** [review]

**Comment by bravo-kernel** [comment]
Much obliged gentlemen, awesome job 💯

**Comment by rchavik** [comment]
Yay.

I'll start rebasing the other one

**Pull request merged**